libvmaf/src/predict.c

```diff
@@ -61,30 +61,143 @@ static int denormalize(const VmafModel *model, double *prediction)
     return 0;
 }
 
+int find_linear_function_parameters(Point p1, Point p2, double *alpha, double *beta) {
 
-static int transform(const VmafModel *model, double *prediction,
+    if (!(p1.x <= p2.x && p1.y <= p2.y))
+        return -EINVAL;  // first_point coordinates need to be smaller or equal to second_point coordinates
+
+    if (p2.x - p1.x == 0 || p2.y - p1.y == 0) {
+        if (!(p1.x == p2.x && p1.y == p2.y))
+            return -EINVAL;  // first_point and second_point cannot lie on a horizontal or vertical line
+        *alpha = 1.0;  // both points are the same
+        *beta = 0.0;
+    }
+    else if (p1.x == 0) {
+        *beta = p1.y;
+        *alpha = (p2.y - *beta) / p2.x;
+    }
+    else {
+        *beta = (p2.y * (p1.y - p1.x)) / (p2.x - p1.x);
+        *alpha = (p1.y - *beta) / p1.x;
+    }
+
+    return 0;
+}
+
+int piecewise_linear_mapping(double x, Point *knots, unsigned n_knots, double *y) {
+
+    double slope, offset;
+
+    if (n_knots <= 1)
+        return EINVAL;
+    unsigned n_seg = n_knots - 1;
+
+    *y = 0.0;
+
+    // construct the function
+    for (unsigned idx = 0; idx < n_seg; idx++) {
```

Comment:
This loop continues even if y has already been found. We may consider break after *y is assigned.

Reply:
Let me keep this for now. Feel free to update the code, but make sure the C and Python version are consistent.